flex_model/core/wrapper.py

```diff
@@ -133,46 +181,162 @@ def __init__(
+            hook_type in self._hook_type_to_pt_attr
+        ), "Invalid hook type provided: {hook_type}"
+
+        # Set private HF state.
+        assert hook_function._shared_state is None
+        assert hook_function._hook_type is None
+        hook_function._shared_state = self._shared_state
+        hook_function._hook_type = hook_type
+
+        # Pass to registration impl.
+        self._register_hook_impl(hook_function)
+
+    def register_forward_hook(self, hook_function: HookFunction, lazy=False) -> None:
+        self._hook_registration_prologue(hook_function, "forward", lazy)
+
+    def register_full_backward_hook(
+        self, hook_function: HookFunction, lazy=False
+    ) -> None:
+        self._hook_registration_prologue(hook_function, "full_backward", lazy)
+
+    def register_hook(self, hook_function: HookFunction, lazy=False) -> None:
+        self._hook_registration_prologue(hook_function, "tensor", lazy)
+
+    def register_forward_pre_hook(
+        self, hook_function: HookFunction, lazy=False
+    ) -> None:
+        self._hook_registration_prologue(hook_function, "forward_pre", lazy)
+
+    def register_full_backward_pre_hook(
+        self, hook_function: HookFunction, lazy=False
+    ) -> None:
+        self._hook_registration_prologue(hook_function, "backward_pre", lazy)
+
+    def _enable_group_hooks(self, group: str):
+        for m, hook_fn_to_handle in self.hook_functions[group].items():
+            for hook_fn, handle in hook_fn_to_handle.items():
+                if handle is None:
+                    self._register_hook_impl(m, hook_fn)
+
+    def _disable_group_hooks(self, group: str):
+        for m, hook_fn_to_handle in self.hook_functions[group].items():
+            for hook_fn, handle in hook_fn_to_handle.items():
+                if handle is not None:
+                    handle.remove()
+                    self.hook_functions[group][m][hook_fn] = None
+
+    def _flush_pipeline(self) -> None:
+        """Gather tensors from all pipeline stages to the first stage."""
+        if (
+            torch.distributed.is_initialized()
+            and dist.in_pipeline_parallel_group()
+            and dist.get_activation_pipeline_parallel_world_size() > 1
+        ):
+            gathered_acts = dist.gather_pipeline_parallel_tensor_dicts(self.output_ptr)
+
+            # Rank 0 accumulates the activation tensors.
+            if dist.get_activation_pipeline_parallel_rank() == 0:
+                self.output_ptr.update(gathered_acts)
 
-        hook_function._output_ptr = self.output_ptr
-        hook_function.save_ctx = self.save_ctx
-        hook_function.modules = self.trainable_modules
-        self.hook_functions[hook_function.module_name] = hook_function
+            # Other ranks reset their collections for the next forward pass.
+            else:
+                self.output_ptr = {}
+
+    def forward(self, *args, _group: str = "all", **kwargs):
+        self._enable_group_hooks(_group)
```

Comment:
Future `TODO`: Provide an interable of all group names and run the union of them.